File: runtime/go/units/wasi/random_get.go
// requires: memory/init
func (w *WASI) wasi_random_get(bufPtr, length uint32) uint32 {
    b := make([]byte, length)
    if _, err := rand.Read(b); err != nil {
        return wasiIo
    }
    w.memory.init(uint64(bufPtr), b, 0, uint64(length))
    return wasiOk
}

File: runtime/perl/units/wasi/random_get.pl
# requires: memory/init
# /dev/urandom is the core-perl entropy source; the rand() tail is a
# non-cryptographic fallback for hosts without it.
sub wasi_random_get {
    my ($self, $buf_ptr, $len) = @_;
    my $bytes = '';
    if (open(my $fh, '<:raw', '/dev/urandom')) {
        while (length($bytes) < $len) {
            last unless sysread($fh, $bytes, $len - length($bytes), length($bytes));
        }
        close($fh);
    }
    $bytes .= chr(int(rand(256))) while length($bytes) < $len;
    $self->{memory}->init($buf_ptr, $bytes, 0, $len);
    return ERRNO_SUCCESS;
}
